src/resilience/backend/FenixBackend.cpp

```
@@ -0,0 +1,389 @@
+ * notice, this list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright
+ * notice, this list of conditions and the following disclaimer in the
+ * documentation and/or other materials provided with the distribution.
+ *
+ * 3. Neither the name of the Corporation nor the names of the
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY NTESS "AS IS" AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL NTESS OR THE
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+ * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+ * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+ */
+#include "FenixBackend.hpp"
+
+#include <sstream>
+#include <fstream>
+#include <cassert>
+
+#include <fenix.h>
+
+#include <resilience/AutomaticCheckpoint.hpp>
+
+#ifdef KR_ENABLE_TRACING
+#include <resilience/util/Trace.hpp>
+#endif
+
+#include <fenix.h>
+
+// TODO: implement exception class
+//Kokkos::Impl::throw_runtime_exception( out.str() );
+#define FENIX_THROW( msg ) std::cerr << msg << std::endl
+
+namespace KokkosResilience
+{
+  class FenixClient
+  {
+    using serializer_t   = std::function< bool( std::ostream & ) >;
+    using deserializer_t = std::function< bool( std::istream & ) >;
+
+    struct ProtectedMemoryBlock
```

Just use the Registrations directly, instead of `ProtectedMemoryBlock`, and use the context's `get_scratch_buffer` instead of consuming memory per member. Call `Fenix_Data_member_create` every time, and ignore the error returned if you are registering the same ID, so you don't need to track `m_registered`. Then call `Fenix_Data_member_attr_set` to update the pointer each time you (re)store, so you can use a scratch buffer instead of dedicated memory.